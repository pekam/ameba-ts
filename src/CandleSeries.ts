import { Candle } from "./loadData";

export class CandleSeries extends Array<Candle> {

  /**
   * @param candles in chronological order
   */
  constructor(...candles: Candle[]) {
    super(...candles);

    // https://github.com/Microsoft/TypeScript/issues/18035
    Object.setPrototypeOf(this, CandleSeries.prototype);
  }

  /**
   * Returns a new series including only the candles
   * of this series that are in the given time range.
   * 
   * @param from the lower limit as unix timestamp, inclusive
   * @param to the upper limit as unix timestamp, exclusive
   */
  subSeries(from: number, to: number): CandleSeries {
    return new CandleSeries(...this.filter(candle =>
      candle.time.getTime() >= from && candle.time.getTime() < to));
  }

  /**
   * Returns a TimeTraveller for this series, that can be
   * used to simulate how the data is received one candle
   * at a time.
   * 
   * @param startTime the unix time of the last candle to
   * be included in the first iteration
   */
  getTimeTraveller(startTime: number): TimeTraveller {
    return new TimeTraveller(this, startTime);
  }

  /**
   * Gets the last candle in the series.
   */
  get last(): Candle {
    return this[this.length - 1];
  }

  /**
   * Gets the time range of the candles as unix timestamps.
   */
  get range(): { start: number, end: number } {
    const dateRange = this.rangeAsDates;
    return {
      start: dateRange.start.getTime(),
      end: dateRange.end.getTime()
    }
  }

  /**
   * Gets the time range of the candles as Date objects.
   */
  get rangeAsDates(): { start: Date, end: Date } {
    return {
      start: this[0].time,
      end: this[this.length - 1].time
    };
  }

}

/**
 * Allows iterating over a CandleSeries, by expanding a
 * subseries with one more candle, each time when calling
 * the next() function.
 * 
 * This can be used by the backtesting system, to simulate
 * going through the history one candle at a time.
 */
export class TimeTraveller {

  private readonly series: CandleSeries;
  /**
   * Index of the candle that will be the last one included
   * in the next call of next() function.
   */
  private nextIndex: number;

  constructor(series: CandleSeries, startTime: number) {
    this.series = series;
    this.nextIndex = series.findIndex(candle =>
      candle.time.getTime() === startTime);
    if (this.nextIndex < 0) {
      throw new Error('Failed to create TimeTraveller. ' +
        'Could not find a candle with startTime ' + startTime);
    }
  }

  /**
   * Returns the next subseries, including one more candle
   * compared to the previous call of this function.
   */
  next(): CandleSeries {
    if (!this.hasNext()) {
      throw new Error('TimeTraveller index out of bounds.');
    }
    return new CandleSeries(...this.series.slice(0, ++this.nextIndex));
  }

  /**
   * Returns true if the iteration has not finished, and the
   * next() function can still be called succesfully.
   */
  hasNext(): boolean {
    return this.nextIndex >= 0 && this.nextIndex < this.series.length;
  }

}
