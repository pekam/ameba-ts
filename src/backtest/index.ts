export {
  backtest,
  BacktestAsyncArgs,
  backtestSync,
  BacktestSyncArgs,
  CommissionProvider,
} from "./backtest";
export { getPersistedBacktestResult } from "./backtest-persistence";
export {
  BacktestResult,
  BacktestStatistics,
  BacktestSyncResult,
  BacktestSyncStatistics,
} from "./backtest-result";
export { AsyncCandleUpdateProvider } from "./candle-update-provider-async";
export * from "./equity-curve";
export { ProgressHandler } from "./progress-handler";
