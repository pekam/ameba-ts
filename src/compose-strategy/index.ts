export * from "./compose-strategy";
export * from "./entries";
export * from "./exits";
export * from "./filters";
