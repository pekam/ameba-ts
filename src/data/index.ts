export * from "./candle-data-provider";
