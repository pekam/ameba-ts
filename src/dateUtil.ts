export function timestampFromUTC(year: number, month: number,
                                 date?: number, hours?: number,
                                 minutes?: number, seconds?: number, ms?: number) {
    return Math.floor(Date.UTC(
        year, month, date, hours || 0, minutes || 0, seconds || 0, ms || 0)
        / 1000);
}