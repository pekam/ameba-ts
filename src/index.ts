import { loadForexCandles, Candle } from "./loadData";

loadForexCandles({
  // symbol: 'OANDA:EUR_USD',
  symbol: 'FXCM:EUR/USD',
  // symbol: 'FXPRO:1',
  resolution: '60',
  from: Date.UTC(2020, 2, 4),
  to: Date.UTC(2020, 2, 5)
})

  .then(candles => {
    console.log(candles)
  })

  .catch(err => {
    console.error('Failed to get forex candles:\n', err)
  })
