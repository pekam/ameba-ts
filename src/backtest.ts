import { Strategy, TradeState, Order, Transaction, MarketPosition, Trade, Candle } from "./types";
import { CandleSeries, TimeTraveller } from "./CandleSeries";

/**
 * Tests how the given strategy would have performed with
 * the provided historical price data.
 *
 * @param strat the strategy to test
 * @param series data series covering at least the range
 * between 'from' and 'to' arguments, plus X time before
 * to have some history for the first values
 * @param from the start of the time range to test
 * as unix timestamp, inclusive
 * @param to the end of the time range to test
 * as unix timestamp, exclusive, can be dismissed
 * to test until the end of the series
 */
export function backtestStrategy(strat: Strategy, series: CandleSeries,
  from: number, to?: number): Trade[] {

  const tt = series.getTimeTraveller(from, to);

  const initialState: TradeState = {
    series: null,
    entryOrder: null,
    position: null,
    stopLoss: null,
    takeProfit: null,
    transactions: []
  }

  const finalState = nextState(initialState, tt, strat);

  return convertToTrades(finalState.transactions);
}

function nextState(state: TradeState, tt: TimeTraveller,
  strat: Strategy): TradeState {

  if (!tt.hasNext()) {
    return state;
  }

  return nextState(
    applyStrategy(
      handleTakeProfit(
        handleStopLoss(
          handleEntryOrder(
            { ...state, series: tt.next() }
          ))),
      strat),
    tt, strat);
}

function handleEntryOrder(state: TradeState) {
  if (!state.position && state.entryOrder &&
    isOrderFulfilled(state.entryOrder, state.series.last)) {
    const mutations = fulfillEntryOrder(state);
    return { ...state, ...mutations };
  }
  return state;
}

function handleStopLoss(state: TradeState) {
  if (state.position && state.stopLoss) {
    const stopLossOrder: Order = createStopLossOrder(state);
    if (isOrderFulfilled(stopLossOrder, state.series.last)) {
      const mutations = fulfillExitOrder(stopLossOrder, state);
      return { ...state, ...mutations };
    }
  }
  return state;
}

function handleTakeProfit(state: TradeState) {
  if (state.position && state.takeProfit) {
    const takeProfitOrder: Order = createTakeProfitOrder(state);
    if (isOrderFulfilled(takeProfitOrder, state.series.last)) {
      const mutations = fulfillExitOrder(takeProfitOrder, state);
      return { ...state, ...mutations };
    }
  }
  return state;
}

function applyStrategy(state: TradeState, strat: Strategy) {
  const mutations = strat(state);
  if (state.position && mutations.entryOrder) {
    throw new Error('Changing entry order while already ' +
      'in a position is not allowed.');
  }
  return { ...state, ...mutations };
}

function isOrderFulfilled(order: Order, newCandle: Candle): boolean {

  const priceBelowOrder = newCandle.low <= order.price;
  const priceAboveOrder = newCandle.high >= order.price;

  return (!order.sell && order.type === 'limit' && priceBelowOrder)
    || (!order.sell && order.type === 'stop' && priceAboveOrder)
    || (order.sell && order.type === 'limit' && priceAboveOrder)
    || (order.sell && order.type === 'stop' && priceBelowOrder)
}

function fulfillEntryOrder(state: TradeState) {
  const transaction = createTransaction(
    state.entryOrder, state.series.last.time);

  const transactions = state.transactions.concat(transaction);

  const position: MarketPosition =
    state.entryOrder.sell ? 'short' : 'long';

  return { transactions, position }
}

function createTransaction(order: Order, time: number): Transaction {
  return {
    sell: order.sell,
    order,
    price: order.price, // ignoring slippage
    time
  }
}

function createStopLossOrder(state: TradeState): Order {
  return {
    price: state.stopLoss,
    type: 'stop',
    sell: state.position === 'long'
  }
}

function createTakeProfitOrder(state: TradeState): Order {
  return {
    price: state.takeProfit,
    type: 'limit',
    sell: state.position === 'long'
  }
}

function fulfillExitOrder(order: Order, state: TradeState) {
  const transaction: Transaction = createTransaction(
    order, state.series.last.time);
  return {
    transactions: state.transactions.concat(transaction),
    position: null,
    entryOrder: null,
    stopLoss: null,
    takeProfit: null
  }
}

function convertToTrades(transactions: Transaction[]): Trade[] {
  return transactions
    // Every other transaction is an exit.
    // If the last entry didn't close, it's ignored.
    .filter((_, i) => i % 2 === 1)
    .map((exit, i) => {
      const entry = transactions[i * 2];
      const position = entry.sell ? 'short' : 'long';
      const profit = entry.sell ?
        (entry.price - exit.price) / entry.price
        : (exit.price - entry.price) / entry.price;
      return { entry, exit, position, profit }
    });
}
