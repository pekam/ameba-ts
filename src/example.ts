import { backtest } from "./core/backtest";
import { Indicators } from "./core/indicators";
import { createStaker, TradingStrategy, withStaker } from "./core/staker";
import { AssetState } from "./core/types";
import { ftxDataStore } from "./ftx/ftx-data-store";

/**
 * Creates a strategy that buys the breakout of previous candle's high if the
 * price is above a simple moving average (SMA), aims to take a 5% profit and
 * has a stop loss that trails 1% below the price.
 *
 * @param smaPeriod the period to use for the moving average
 */
function exampleStrategy(smaPeriod: number): TradingStrategy {
  // Calculating indicator values can be really time consuming, unless storing
  // the previous values in a closure. Unfortunately this makes the strategy
  // function stateful.
  const indicators = new Indicators({ smaPeriod });

  return (state: AssetState) => {
    const { sma } = indicators.update(state.series);
    const newCandle = state.series[state.series.length - 1];

    if (!sma) {
      return { entryOrder: null };
    }

    if (!state.position) {
      const entryPrice = newCandle.high;
      // Enter only above the moving average
      if (entryPrice < sma) {
        return { entryOrder: null };
      }

      return {
        entryOrder: {
          side: "buy",
          type: "stop",
          price: entryPrice,
        },
        stopLoss: entryPrice * 0.99,
        takeProfit: entryPrice * 1.05,
      };
    } else {
      // Manage the current position
      return {
        // Trailing 1% stop loss
        stopLoss: Math.max(state.stopLoss || 0, newCandle.high * 0.99),
      };
    }
  };
}

/**
 * Returns hourly candlestick data for BTC and ETH cryptos.
 */
async function getCryptoSeries() {
  // Note that ftxDataStore requires FTX API keys in properties.json and MongoDB
  // to be running, as explained in README.
  return {
    BTC: await ftxDataStore.getCandles({
      market: "BTC/USD",
      startDate: "2021-01-01",
      endDate: "2021-12-01",
      resolution: "1h",
    }),
    ETH: await ftxDataStore.getCandles({
      market: "ETH/USD",
      startDate: "2021-01-01",
      endDate: "2021-12-01",
      resolution: "1h",
    }),
  };
}

// Backtest the example strategy with the crypto data
(async () => {
  const result = backtest({
    strategy: withStaker(
      () => exampleStrategy(50),
      // Staker handles position sizing/risk management
      createStaker({
        maxRelativeRisk: 0.01, // Risk 1% of account per trade
        maxRelativeExposure: 1.5, // Use max 50% leverage
        allowFractions: true, // Cryptos allow non-whole-number position sizes
      })
    ),
    series: await getCryptoSeries(),
  });

  console.log(result.stats);
  /*
  {
    initialBalance: 10000,
    endBalance: 32161.101496586634,
    relativeProfit: 2.2161101496586633,
    tradeCount: 2183,
    successRate: 0.37792029317453046,
    buyAndHoldProfit: 3.1247092452187273,
    range: { from: 1609459200, to: 1638313200 }
  }

  222% profit looks awesome, until comparing to the buy-and-hold profit,
  which is 312% on average for these two cryptos. Also, transaction costs
  and slippage (to be implemented) for 2183 trades could very well turn
  this result negative.
  */
})();
