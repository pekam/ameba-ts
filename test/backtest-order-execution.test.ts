import { omit, pipe } from "remeda";
import { AssetState, Candle } from "../src";
import {
  handleOrders,
  OrderHandlerArgs,
} from "../src/backtest/backtest-order-execution";

type TestArgs = Partial<
  Pick<AssetState, "position" | "entryOrder" | "stopLoss" | "takeProfit">
> & { candle: Candle };

function getArgs(args: TestArgs): OrderHandlerArgs {
  const asset: AssetState = {
    symbol: "foo",
    series: [args.candle],
    position: null,
    entryOrder: null,
    stopLoss: null,
    takeProfit: null,
    bufferSize: 100,
    data: {},
    transactions: [],
    trades: [],
    ...omit(args, ["candle"]),
  };
  return {
    asset,
    cash: 100,
    commissionProvider: () => 1,
  };
}

function testHandleOrders(args: TestArgs) {
  const result = pipe(args, getArgs, handleOrders, (r) => ({
    ...r,
    asset: omit(r.asset, ["bufferSize", "data", "series", "symbol"]),
  }));
  expect(result).toMatchSnapshot();
}

const greenCandle: Candle = {
  open: 50,
  close: 60,
  low: 40,
  high: 70,
  volume: 100,
  time: 1,
};

const redCandle: Candle = {
  open: 50,
  close: 40,
  low: 30,
  high: 60,
  volume: 100,
  time: 1,
};

// green candle, long position

it("green candle: should fill entry at bottom tail, skip sl below tail, fill tp at high", () => {
  testHandleOrders({
    candle: greenCandle,
    entryOrder: { side: "buy", type: "limit", size: 1, price: 40 },
    stopLoss: 30,
    takeProfit: 70,
  });
});

it("green candle: should fill entry at bottom tail, fill sl at bottom tail, skip tp at high", () => {
  testHandleOrders({
    candle: greenCandle,
    entryOrder: { side: "buy", type: "limit", size: 1, price: 45 },
    stopLoss: 40,
    takeProfit: 70,
  });
});

it("green candle: should fill entry at body, skip sl in body below entry, fill tp at high", () => {
  testHandleOrders({
    candle: greenCandle,
    entryOrder: { side: "buy", type: "stop", size: 1, price: 55 },
    stopLoss: 52,
    takeProfit: 70,
  });
});

it("green candle: should fill entry and sl immediately", () => {
  testHandleOrders({
    candle: greenCandle,
    entryOrder: { side: "buy", type: "limit", size: 1, price: 55 },
    stopLoss: 52,
  });
});

// green candle, short position

it("green candle: should fill entry", () => {
  testHandleOrders({
    candle: greenCandle,
    entryOrder: { side: "sell", type: "limit", size: 1, price: 55 },
  });
});

it("green candle: should fill short entry at body, skip tp in body below entry, fill sl at high", () => {
  testHandleOrders({
    candle: greenCandle,
    entryOrder: { side: "sell", type: "limit", size: 1, price: 55 },
    stopLoss: 70,
    takeProfit: 52,
  });
});

// red candle, short position

it("red candle: should fill entry at top tail, skip sl above tail, fill tp at low", () => {
  testHandleOrders({
    candle: redCandle,
    entryOrder: { side: "sell", type: "limit", size: 1, price: 60 },
    stopLoss: 70,
    takeProfit: 30,
  });
});

it("red candle: should fill entry at top tail, fill sl at top tail, skip tp at low", () => {
  testHandleOrders({
    candle: redCandle,
    entryOrder: { side: "sell", type: "limit", size: 1, price: 55 },
    stopLoss: 60,
    takeProfit: 30,
  });
});

it("red candle: should fill entry at body, skip sl in body above entry, fill tp at low", () => {
  testHandleOrders({
    candle: redCandle,
    entryOrder: { side: "sell", type: "stop", size: 1, price: 45 },
    stopLoss: 48,
    takeProfit: 30,
  });
});

it("red candle: should fill entry and sl immediately", () => {
  testHandleOrders({
    candle: redCandle,
    entryOrder: { side: "sell", type: "limit", size: 1, price: 45 },
    stopLoss: 48,
  });
});

// red candle, long position

it("red candle: should fill entry", () => {
  testHandleOrders({
    candle: redCandle,
    entryOrder: { side: "buy", type: "limit", size: 1, price: 45 },
  });
});

it("red candle: should fill long entry at body, skip tp in body above entry, fill sl at low", () => {
  testHandleOrders({
    candle: redCandle,
    entryOrder: { side: "buy", type: "limit", size: 1, price: 45 },
    stopLoss: 30,
    takeProfit: 48,
  });
});
