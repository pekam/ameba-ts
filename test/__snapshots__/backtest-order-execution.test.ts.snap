// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`green candle: should fill entry 1`] = `
{
  "asset": {
    "entryOrder": {
      "price": 55,
      "side": "sell",
      "size": 1,
      "type": "limit",
    },
    "position": {
      "side": "short",
      "size": 1,
    },
    "stopLoss": null,
    "takeProfit": null,
    "trades": [],
    "transactions": [
      {
        "commission": 1,
        "price": 55,
        "side": "sell",
        "size": 1,
        "time": 1,
      },
    ],
  },
  "cash": 154,
}
`;

exports[`green candle: should fill entry and sl immediately 1`] = `
{
  "asset": {
    "entryOrder": null,
    "position": null,
    "stopLoss": null,
    "takeProfit": null,
    "trades": [
      {
        "absoluteProfit": -2,
        "entry": {
          "commission": 1,
          "price": 50,
          "side": "buy",
          "size": 1,
          "time": 1,
        },
        "exit": {
          "commission": 1,
          "price": 50,
          "side": "sell",
          "size": 1,
          "time": 1,
        },
        "position": {
          "side": "long",
          "size": 1,
        },
        "relativeProfit": -0.04,
        "symbol": "foo",
      },
    ],
    "transactions": [
      {
        "commission": 1,
        "price": 50,
        "side": "buy",
        "size": 1,
        "time": 1,
      },
      {
        "commission": 1,
        "price": 50,
        "side": "sell",
        "size": 1,
        "time": 1,
      },
    ],
  },
  "cash": 98,
}
`;

exports[`green candle: should fill entry at body, skip sl in body below entry, fill tp at high 1`] = `
{
  "asset": {
    "entryOrder": null,
    "position": null,
    "stopLoss": null,
    "takeProfit": null,
    "trades": [
      {
        "absoluteProfit": 13,
        "entry": {
          "commission": 1,
          "price": 55,
          "side": "buy",
          "size": 1,
          "time": 1,
        },
        "exit": {
          "commission": 1,
          "price": 70,
          "side": "sell",
          "size": 1,
          "time": 1,
        },
        "position": {
          "side": "long",
          "size": 1,
        },
        "relativeProfit": 0.23636363636363636,
        "symbol": "foo",
      },
    ],
    "transactions": [
      {
        "commission": 1,
        "price": 55,
        "side": "buy",
        "size": 1,
        "time": 1,
      },
      {
        "commission": 1,
        "price": 70,
        "side": "sell",
        "size": 1,
        "time": 1,
      },
    ],
  },
  "cash": 113,
}
`;

exports[`green candle: should fill entry at bottom tail, fill sl at bottom tail, skip tp at high 1`] = `
{
  "asset": {
    "entryOrder": null,
    "position": null,
    "stopLoss": null,
    "takeProfit": null,
    "trades": [
      {
        "absoluteProfit": -7,
        "entry": {
          "commission": 1,
          "price": 45,
          "side": "buy",
          "size": 1,
          "time": 1,
        },
        "exit": {
          "commission": 1,
          "price": 40,
          "side": "sell",
          "size": 1,
          "time": 1,
        },
        "position": {
          "side": "long",
          "size": 1,
        },
        "relativeProfit": -0.15555555555555556,
        "symbol": "foo",
      },
    ],
    "transactions": [
      {
        "commission": 1,
        "price": 45,
        "side": "buy",
        "size": 1,
        "time": 1,
      },
      {
        "commission": 1,
        "price": 40,
        "side": "sell",
        "size": 1,
        "time": 1,
      },
    ],
  },
  "cash": 93,
}
`;

exports[`green candle: should fill entry at bottom tail, skip sl below tail, fill tp at high 1`] = `
{
  "asset": {
    "entryOrder": null,
    "position": null,
    "stopLoss": null,
    "takeProfit": null,
    "trades": [
      {
        "absoluteProfit": 28,
        "entry": {
          "commission": 1,
          "price": 40,
          "side": "buy",
          "size": 1,
          "time": 1,
        },
        "exit": {
          "commission": 1,
          "price": 70,
          "side": "sell",
          "size": 1,
          "time": 1,
        },
        "position": {
          "side": "long",
          "size": 1,
        },
        "relativeProfit": 0.7,
        "symbol": "foo",
      },
    ],
    "transactions": [
      {
        "commission": 1,
        "price": 40,
        "side": "buy",
        "size": 1,
        "time": 1,
      },
      {
        "commission": 1,
        "price": 70,
        "side": "sell",
        "size": 1,
        "time": 1,
      },
    ],
  },
  "cash": 128,
}
`;

exports[`green candle: should fill short entry at body, skip tp in body below entry, fill sl at high 1`] = `
{
  "asset": {
    "entryOrder": null,
    "position": null,
    "stopLoss": null,
    "takeProfit": null,
    "trades": [
      {
        "absoluteProfit": -17,
        "entry": {
          "commission": 1,
          "price": 55,
          "side": "sell",
          "size": 1,
          "time": 1,
        },
        "exit": {
          "commission": 1,
          "price": 70,
          "side": "buy",
          "size": 1,
          "time": 1,
        },
        "position": {
          "side": "short",
          "size": 1,
        },
        "relativeProfit": -0.3090909090909091,
        "symbol": "foo",
      },
    ],
    "transactions": [
      {
        "commission": 1,
        "price": 55,
        "side": "sell",
        "size": 1,
        "time": 1,
      },
      {
        "commission": 1,
        "price": 70,
        "side": "buy",
        "size": 1,
        "time": 1,
      },
    ],
  },
  "cash": 83,
}
`;

exports[`red candle: should fill entry 1`] = `
{
  "asset": {
    "entryOrder": {
      "price": 45,
      "side": "buy",
      "size": 1,
      "type": "limit",
    },
    "position": {
      "side": "long",
      "size": 1,
    },
    "stopLoss": null,
    "takeProfit": null,
    "trades": [],
    "transactions": [
      {
        "commission": 1,
        "price": 45,
        "side": "buy",
        "size": 1,
        "time": 1,
      },
    ],
  },
  "cash": 54,
}
`;

exports[`red candle: should fill entry and sl immediately 1`] = `
{
  "asset": {
    "entryOrder": null,
    "position": null,
    "stopLoss": null,
    "takeProfit": null,
    "trades": [
      {
        "absoluteProfit": -2,
        "entry": {
          "commission": 1,
          "price": 50,
          "side": "sell",
          "size": 1,
          "time": 1,
        },
        "exit": {
          "commission": 1,
          "price": 50,
          "side": "buy",
          "size": 1,
          "time": 1,
        },
        "position": {
          "side": "short",
          "size": 1,
        },
        "relativeProfit": -0.04,
        "symbol": "foo",
      },
    ],
    "transactions": [
      {
        "commission": 1,
        "price": 50,
        "side": "sell",
        "size": 1,
        "time": 1,
      },
      {
        "commission": 1,
        "price": 50,
        "side": "buy",
        "size": 1,
        "time": 1,
      },
    ],
  },
  "cash": 98,
}
`;

exports[`red candle: should fill entry at body, skip sl in body above entry, fill tp at low 1`] = `
{
  "asset": {
    "entryOrder": null,
    "position": null,
    "stopLoss": null,
    "takeProfit": null,
    "trades": [
      {
        "absoluteProfit": 13,
        "entry": {
          "commission": 1,
          "price": 45,
          "side": "sell",
          "size": 1,
          "time": 1,
        },
        "exit": {
          "commission": 1,
          "price": 30,
          "side": "buy",
          "size": 1,
          "time": 1,
        },
        "position": {
          "side": "short",
          "size": 1,
        },
        "relativeProfit": 0.28888888888888886,
        "symbol": "foo",
      },
    ],
    "transactions": [
      {
        "commission": 1,
        "price": 45,
        "side": "sell",
        "size": 1,
        "time": 1,
      },
      {
        "commission": 1,
        "price": 30,
        "side": "buy",
        "size": 1,
        "time": 1,
      },
    ],
  },
  "cash": 113,
}
`;

exports[`red candle: should fill entry at top tail, fill sl at top tail, skip tp at low 1`] = `
{
  "asset": {
    "entryOrder": null,
    "position": null,
    "stopLoss": null,
    "takeProfit": null,
    "trades": [
      {
        "absoluteProfit": -7,
        "entry": {
          "commission": 1,
          "price": 55,
          "side": "sell",
          "size": 1,
          "time": 1,
        },
        "exit": {
          "commission": 1,
          "price": 60,
          "side": "buy",
          "size": 1,
          "time": 1,
        },
        "position": {
          "side": "short",
          "size": 1,
        },
        "relativeProfit": -0.12727272727272726,
        "symbol": "foo",
      },
    ],
    "transactions": [
      {
        "commission": 1,
        "price": 55,
        "side": "sell",
        "size": 1,
        "time": 1,
      },
      {
        "commission": 1,
        "price": 60,
        "side": "buy",
        "size": 1,
        "time": 1,
      },
    ],
  },
  "cash": 93,
}
`;

exports[`red candle: should fill entry at top tail, skip sl above tail, fill tp at low 1`] = `
{
  "asset": {
    "entryOrder": null,
    "position": null,
    "stopLoss": null,
    "takeProfit": null,
    "trades": [
      {
        "absoluteProfit": 28,
        "entry": {
          "commission": 1,
          "price": 60,
          "side": "sell",
          "size": 1,
          "time": 1,
        },
        "exit": {
          "commission": 1,
          "price": 30,
          "side": "buy",
          "size": 1,
          "time": 1,
        },
        "position": {
          "side": "short",
          "size": 1,
        },
        "relativeProfit": 0.4666666666666667,
        "symbol": "foo",
      },
    ],
    "transactions": [
      {
        "commission": 1,
        "price": 60,
        "side": "sell",
        "size": 1,
        "time": 1,
      },
      {
        "commission": 1,
        "price": 30,
        "side": "buy",
        "size": 1,
        "time": 1,
      },
    ],
  },
  "cash": 128,
}
`;

exports[`red candle: should fill long entry at body, skip tp in body above entry, fill sl at low 1`] = `
{
  "asset": {
    "entryOrder": null,
    "position": null,
    "stopLoss": null,
    "takeProfit": null,
    "trades": [
      {
        "absoluteProfit": -17,
        "entry": {
          "commission": 1,
          "price": 45,
          "side": "buy",
          "size": 1,
          "time": 1,
        },
        "exit": {
          "commission": 1,
          "price": 30,
          "side": "sell",
          "size": 1,
          "time": 1,
        },
        "position": {
          "side": "long",
          "size": 1,
        },
        "relativeProfit": -0.37777777777777777,
        "symbol": "foo",
      },
    ],
    "transactions": [
      {
        "commission": 1,
        "price": 45,
        "side": "buy",
        "size": 1,
        "time": 1,
      },
      {
        "commission": 1,
        "price": 30,
        "side": "sell",
        "size": 1,
        "time": 1,
      },
    ],
  },
  "cash": 83,
}
`;
